Source/Dafny/Compilers/Compiler-Csharp.cs

```diff
@@ -642,14 +642,16 @@ bool ContainsTyVar(TypeParameter tp, Type ty)
     }
 
     private void CompileDatatypeDestructorsAndAddToInterface(DatatypeDecl dt, ConcreteSyntaxTree wr,
-      ConcreteSyntaxTree interfaceTree, string DtT_TypeArgs) {
+        ConcreteSyntaxTree interfaceTree, string DtT_TypeArgs) {
       foreach (var ctor in dt.Ctors) {
-        foreach (var dtor in ctor.Destructors) {
+        foreach (var (dtor, index) in ctor.Destructors.Select((v, i) => (v, i))) {
           if (dtor.EnclosingCtors[0] == ctor) {
             var arg = dtor.CorrespondingFormals[0];
-            if (!arg.IsGhost && arg.HasName) {
+            if (!arg.IsGhost) {
+              var name = arg.HasName ? arg.CompileName : FormalName(arg, index);
```

Comment:
How come you can't write `var name = FormalName(arg, index);` here? The `HasName` check is already in FormalName. Is the extra `IdProtect` call a problem here?

Reply:
I believe that's exactly right.

Reply:
We also don't want an index when there is a name.

Reply:
You can't have something like `dtor_@this`.